pkg/ingester/client/custom.go

```diff
@@ -9,10 +9,51 @@ import (
 	"github.com/pkg/errors"
 	"github.com/prometheus/prometheus/tsdb/chunkenc"
 	"github.com/prometheus/prometheus/tsdb/chunks"
+	"google.golang.org/grpc/mem"
 
+	"github.com/grafana/mimir/pkg/mimirpb"
 	"github.com/grafana/mimir/pkg/storage/chunk"
 )
 
+var _ mimirpb.BufferHolder = &QueryResponse{}
+
+func (m *QueryResponse) SetBuffer(buf mem.Buffer) {
+	m.buffer = buf
+}
+
+func (m *QueryResponse) FreeBuffer() {
+	if m.buffer != nil {
+		m.buffer.Free()
+		m.buffer = nil
+	}
+}
+
+var _ mimirpb.BufferHolder = &QueryStreamResponse{}
+
+func (m *QueryStreamResponse) SetBuffer(buf mem.Buffer) {
+	m.buffer = buf
+}
+
+func (m *QueryStreamResponse) FreeBuffer() {
+	if m.buffer != nil {
+		m.buffer.Free()
+		m.buffer = nil
+	}
+}
+
+var _ mimirpb.BufferHolder = &ExemplarQueryResponse{}
+
+func (m *ExemplarQueryResponse) SetBuffer(buf mem.Buffer) {
+	m.buffer = buf
+}
+
+func (m *ExemplarQueryResponse) FreeBuffer() {
+	if m.buffer != nil {
+		m.buffer.Free()
+		m.buffer = nil
+	}
+}
```

Comment:
All the BufferHolder implementation are the same?

I think _this_ might be a good candidate for embedding. 

What if we defined `BufferHolder` as the implementation, not an interface, and just embedded it in each one of the messages?

Reply:
Could we leave this to another PR? It would be a pain to fix up the *.pb.expdiff files to do this, and there are relatively few types affected. I would instead make this improvement in the future if more types would need to be `BufferHolder`s.

Reply:
I think that embedding a type reduces the amount of code and complexity, and we shouldn't merge something more complex just because we don't have a proper workflow to update the expdiff files.

Maybe this is a good opportunity to create a `make` target that would properly update the expdiff files? I tried following the instructions that you left in our internal weekly meeting, and the results I got don't look good: https://github.com/grafana/mimir/compare/arve/upgrade-grpc...arve/upgrade-grpc-keep-buffer-reference

Reply:
@colega OK, I revised to use an embedded `BufferHolder` from the *.pb.go files, PTAL.